### PR TITLE
Expose methods needed by Ruby FFI using UPB_API

### PR DESCRIPTION
**Pull request by JasonLunn**
Once merged, the changes to ~`upb-ruby.h`~ `glue.c` in https://github.com/protocolbuffers/protobuf/pull/11483 can be reverted without breaking the FFI feature branch.

### CODE REVIEW

File: upb/collections/map.h [hunk 1]
```diff
@@ -120,16 +120,18 @@ UPB_API void upb_Map_SetEntryValue(upb_Map* map, size_t iter,
  */
 
 // Advances to the next entry. Returns false if no more entries are present.
-bool upb_MapIterator_Next(const upb_Map* map, size_t* iter);
+UPB_API_INLINE bool upb_MapIterator_Next(const upb_Map* map, size_t* iter);
```
Comment (haberman) [inline upb/collections/map.h:123]: `UPB_API_INLINE` is intended for inline functions. For non-inline functions, we should use `UPB_API`.

File: upb/hash/common.h [hunk 2]
```diff
@@ -189,7 +189,7 @@ UPB_INLINE size_t upb_table_size(const upb_table* t) {
 
 UPB_INLINE bool upb_tabent_isempty(const upb_tabent* e) { return e->key == 0; }
 
-uint32_t _upb_Hash(const void* p, size_t n, uint64_t seed);
+UPB_API_INLINE uint32_t _upb_Hash(const void* p, size_t n, uint64_t seed);
```
Comment (haberman) [inline upb/hash/common.h:192]: Why is this needed? This is an implementation detail inside upb.

**Review by haberman — CHANGES_REQUESTED** [review]

File: upb/hash/common.h (hunk 2 above)
Comment (JasonLunn) [reply upb/hash/common.h:192]: `Message_Hash` (from `message.c`) and `Msgval_GetHash` (from `convert.c`) call it directly, so FFI needs to be able to do the same.

File: upb/collections/map.h (hunk 1 above)
Comment (JasonLunn) [reply upb/collections/map.h]: Done

File: upb/hash/common.h (hunk 2 above)
Comment (JasonLunn) [reply upb/hash/common.h]: Reverted

**Review by haberman — APPROVED** [review]
